File: templates/components/multiagent/typescript/workflow/tools.ts
import { HandlerContext } from "@llamaindex/workflow";
import {
  BaseToolWithCall,
  callTool,
  ChatMessage,
  ChatResponse,
  ChatResponseChunk,
  PartialToolCall,
  QueryEngineTool,
  ToolCall,
  ToolCallLLM,
  ToolCallLLMMessageOptions,
} from "llamaindex";
import crypto from "node:crypto";
import { getDataSource } from "../engine";
import { createQueryEngineTool } from "../engine/tools/query-engine";
import { AgentRunEvent } from "./type";

export const getQueryEngineTool = async (): Promise<QueryEngineTool | null> => {
  const index = await getDataSource();

  if (!index) {
    return null;
  }

  return createQueryEngineTool(index);
};

/**
 * Call multiple tools and return the tool messages
 */
export const callTools = async <T>({
  tools,
  toolCalls,
  ctx,
  agentName,
  writeEvent = true,
}: {
  toolCalls: ToolCall[];
  tools: BaseToolWithCall[];
  ctx: HandlerContext<T>;
  agentName: string;
  writeEvent?: boolean;
}): Promise<ChatMessage[]> => {
  const toolMsgs: ChatMessage[] = [];
  if (toolCalls.length === 0) {
    return toolMsgs;
  }
  if (toolCalls.length === 1) {
    const tool = tools.find((tool) => tool.metadata.name === toolCalls[0].name);
    if (!tool) {
      throw new Error(`Tool ${toolCalls[0].name} not found`);
    }
    return [
      await callSingleTool(
        tool,
        toolCalls[0],
        writeEvent
          ? (msg: string) => {
              ctx.sendEvent(
                new AgentRunEvent({
                  agent: agentName,
                  text: msg,
                  type: "text",
                }),
              );
            }
          : undefined,
      ),
    ];
  }
  // Multiple tool calls, show events in progress
  const progressId = crypto.randomUUID();
  const totalSteps = toolCalls.length;
  let currentStep = 0;
  for (const toolCall of toolCalls) {
    const tool = tools.find((tool) => tool.metadata.name === toolCall.name);
    if (!tool) {
      throw new Error(`Tool ${toolCall.name} not found`);
    }
    const toolMsg = await callSingleTool(tool, toolCall, (msg: string) => {
      ctx.sendEvent(
        new AgentRunEvent({
          agent: agentName,
          text: msg,
          type: "progress",
          data: {
            id: progressId,
            total: totalSteps,
            current: currentStep,
          },
        }),
      );
      currentStep++;
    });
    toolMsgs.push(toolMsg);
  }
  return toolMsgs;
};

export const callSingleTool = async (
  tool: BaseToolWithCall,
  toolCall: ToolCall,
  eventEmitter?: (msg: string) => void,
): Promise<ChatMessage> => {
  if (eventEmitter) {
    eventEmitter(
      `Calling tool ${toolCall.name} with input: ${JSON.stringify(toolCall.input)}`,
    );
  }

  const toolOutput = await callTool(tool, toolCall, {
    log: () => {},
    error: (...args: unknown[]) => {
      console.error(`Tool ${toolCall.name} got error:`, ...args);
      if (eventEmitter) {
        eventEmitter(`Tool ${toolCall.name} got error: ${args.join(" ")}`);
      }
      return {
        content: JSON.stringify({
          error: args.join(" "),
        }),
        role: "user",
        options: {
          toolResult: {
            id: toolCall.id,
            result: JSON.stringify({
              error: args.join(" "),
            }),
            isError: true,
          },
        },
      };
    },
    warn: () => {},
  });

  return {
    content: JSON.stringify(toolOutput.output),
    role: "user",
    options: {
      toolResult: {
        result: toolOutput.output,
        isError: toolOutput.isError,
        id: toolCall.id,
      },
    },
  };
};

class ChatWithToolsResponse {
  toolCalls: ToolCall[];
  toolCallMessage?: ChatMessage;
  responseGenerator?: AsyncGenerator<ChatResponseChunk>;

  constructor(options: {
    toolCalls: ToolCall[];
    toolCallMessage?: ChatMessage;
    responseGenerator?: AsyncGenerator<ChatResponseChunk>;
  }) {
    this.toolCalls = options.toolCalls;
    this.toolCallMessage = options.toolCallMessage;
    this.responseGenerator = options.responseGenerator;
  }

  hasMultipleTools() {
    const uniqueToolNames = new Set(this.getToolNames());
    return uniqueToolNames.size > 1;
  }

  hasToolCall() {
    return this.toolCalls.length > 0;
  }

  getToolNames() {
    return this.toolCalls.map((toolCall) => toolCall.name);
  }

  async asFullResponse(): Promise<ChatMessage> {
    if (!this.responseGenerator) {
      throw new Error("No response generator");
    }
    let fullResponse = "";
    for await (const chunk of this.responseGenerator) {
      fullResponse += chunk.delta;
    }
    return {
      role: "assistant",
      content: fullResponse,
    };
  }
}

export const chatWithTools = async (
  llm: ToolCallLLM,
  tools: BaseToolWithCall[],
  messages: ChatMessage[],
): Promise<ChatWithToolsResponse> => {
  const responseGenerator = async function* (): AsyncGenerator<
    boolean | ChatResponseChunk,
    void,
    unknown
  > {
    const responseStream = await llm.chat({ messages, tools, stream: true });

    let fullResponse = null;
    let yieldedIndicator = false;
    const toolCallMap = new Map();
    for await (const chunk of responseStream) {
      const hasToolCalls = chunk.options && "toolCall" in chunk.options;
      if (!hasToolCalls) {
        if (!yieldedIndicator) {
          yield false;
          yieldedIndicator = true;
        }
        yield chunk;
      } else if (!yieldedIndicator) {
        yield true;
        yieldedIndicator = true;
      }

      if (chunk.options && "toolCall" in chunk.options) {
        for (const toolCall of chunk.options.toolCall as PartialToolCall[]) {
          if (toolCall.id) {
            toolCallMap.set(toolCall.id, toolCall);
          }
        }
      }

      if (
        hasToolCalls &&
        (chunk.raw as any)?.choices?.[0]?.finish_reason !== null
      ) {
        // Update the fullResponse with the tool calls
        const toolCalls = Array.from(toolCallMap.values());
        fullResponse = {
          ...chunk,
          options: {
            ...chunk.options,
            toolCall: toolCalls,
          },
        };
      }
    }

    if (fullResponse) {
      yield fullResponse;
    }
  };

  const generator = responseGenerator();
  const isToolCall = await generator.next();

  if (isToolCall.value) {
    // If it's a tool call, we need to wait for the full response
    let fullResponse = null;
    for await (const chunk of generator) {
      fullResponse = chunk;
    }

    if (fullResponse) {
      const responseChunk = fullResponse as ChatResponseChunk;
      const toolCalls = getToolCallsFromResponse(responseChunk);
      return new ChatWithToolsResponse({
        toolCalls,
        toolCallMessage: {
          options: responseChunk.options,
          role: "assistant",
          content: "",
        },
      });
    } else {
      throw new Error("Cannot get tool calls from response");
    }
  }

  return new ChatWithToolsResponse({
    toolCalls: [],
    responseGenerator: generator as AsyncGenerator<ChatResponseChunk>,
  });
};

export const getToolCallsFromResponse = (
  response:
    | ChatResponse<ToolCallLLMMessageOptions>
    | ChatResponseChunk<ToolCallLLMMessageOptions>,
): ToolCall[] => {
  let options;

  if ("message" in response) {
    options = response.message.options;
  } else {
    options = response.options;
  }

  if (options && "toolCall" in options) {
    return options.toolCall as ToolCall[];
  }
  return [];
};

File: templates/components/multiagent/python/app/workflows/tools.py
import uuid
from abc import ABC, abstractmethod
from typing import Any, AsyncGenerator, Callable

from app.workflows.events import AgentRunEvent, AgentRunEventType
from llama_index.core.base.llms.types import ChatMessage, ChatResponse, MessageRole
from llama_index.core.llms.function_calling import FunctionCallingLLM
from llama_index.core.tools import (
    BaseTool,
    FunctionTool,
    ToolMetadata,
    ToolOutput,
    ToolSelection,
)
from llama_index.core.workflow import Context
from pydantic import BaseModel, ConfigDict


class ContextAwareTool(FunctionTool, ABC):
    @abstractmethod
    async def acall(self, ctx: Context, input: Any) -> ToolOutput:  # type: ignore
        pass


class ResponseGenerator(BaseModel):
    """
    A response generator from chat_with_tools.
    """

    generator: AsyncGenerator[ChatResponse | None, None]

    model_config = ConfigDict(arbitrary_types_allowed=True)


class ToolCallResponse(BaseModel):
    """
    A tool call response from chat_with_tools.
    """

    tool_calls: list[ToolSelection]
    tool_call_message: ChatMessage


async def workflow_step_as_tool(step: Callable) -> FunctionTool:
    """
    Construct a step as a tool to passing to the agent.
    """
    step_description = step.__doc__ or step.__name__
    tool = FunctionTool(
        fn=step,
        metadata=ToolMetadata(
            name=step.__name__,
            description=step_description,
        ),
    )
    return tool


async def call_tools(
    ctx: Context,
    agent_name: str,
    tools: list[BaseTool],
    tool_calls: list[ToolSelection],
    emit_agent_events: bool = True,
) -> list[ChatMessage]:
    tools_by_name = {tool.metadata.get_name(): tool for tool in tools}
    tool_msgs: list[ChatMessage] = []

    # call tools -- safely!
    # If there are multiple tool calls, show the progress
    progress_id = str(uuid.uuid4())
    total_steps = len(tool_calls)
    show_progress = total_steps > 1
    if show_progress and emit_agent_events:
        ctx.write_event_to_stream(
            AgentRunEvent(
                name=agent_name,
                msg=f"Making {total_steps} tool calls",
            )
        )
    for i, tool_call in enumerate(tool_calls):
        tool = tools_by_name.get(tool_call.tool_name)
        additional_kwargs = {
            "tool_call_id": tool_call.tool_id,
            "name": tool.metadata.get_name(),
        }
        if not tool:
            tool_msgs.append(
                ChatMessage(
                    role=MessageRole.ASSISTANT,
                    content=f"Tool {tool_call.tool_name} does not exist",
                    additional_kwargs=additional_kwargs,
                )
            )
            continue
        try:
            if show_progress and emit_agent_events:
                ctx.write_event_to_stream(
                    AgentRunEvent(
                        name=agent_name,
                        msg=f"Calling tool {tool_call.tool_name}, {tool_call.tool_kwargs}",
                        event_type=AgentRunEventType.PROGRESS,
                        data={
                            "id": progress_id,
                            "total": total_steps,
                            "current": i,
                        },
                    )
                )
            else:
                ctx.write_event_to_stream(
                    AgentRunEvent(
                        name=agent_name,
                        msg=f"Calling tool {tool_call.tool_name}, {str(tool_call.tool_kwargs)}",
                    )
                )
            if isinstance(tool, ContextAwareTool):
                if ctx is None:
                    raise ValueError("Context is required for context aware tool")
                # inject context for calling an context aware tool
                response = await tool.acall(ctx=ctx, **tool_call.tool_kwargs)
            else:
                response = await tool.acall(**tool_call.tool_kwargs)  # type: ignore
            tool_msgs.append(
                ChatMessage(
                    role=MessageRole.TOOL,
                    content=str(response.raw_output),
                    additional_kwargs=additional_kwargs,
                )
            )
        except Exception as e:
            # Print trace back here
            tool_msg = ChatMessage(
                role=MessageRole.TOOL,
                content=f"Error: {str(e)}",
                additional_kwargs=additional_kwargs,
            )
            ctx.write_event_to_stream(
                AgentRunEvent(
                    name=agent_name,
                    msg=f"Error in tool {tool_call.tool_name}: {str(e)}",
                )
            )
            tool_msgs.append(tool_msg)
    if show_progress and emit_agent_events:
        ctx.write_event_to_stream(
            AgentRunEvent(
                name=agent_name,
                msg="Task finished",
            )
        )

    return tool_msgs


async def chat_with_tools(  # type: ignore
    llm: FunctionCallingLLM,
    tools: list[BaseTool],
    chat_history: list[ChatMessage],
) -> ToolCallResponse | ResponseGenerator:
    """
    Request LLM to call tools or not.
    This function doesn't change the memory.
    """
    generator = _tool_call_generator(llm, tools, chat_history)
    is_tool_call = await generator.__anext__()
    if is_tool_call:
        # Last chunk is the full response
        # Wait for the last chunk
        full_response = None
        async for chunk in generator:
            full_response = chunk
        assert isinstance(full_response, ChatResponse)
        return ToolCallResponse(
            tool_calls=llm.get_tool_calls_from_response(full_response),
            tool_call_message=full_response.message,
        )
    else:
        return ResponseGenerator(generator=generator)


async def _tool_call_generator(
    llm: FunctionCallingLLM,
    tools: list[BaseTool],
    chat_history: list[ChatMessage],
) -> AsyncGenerator[ChatResponse | bool, None]:
    response_stream = await llm.astream_chat_with_tools(
        tools,
        chat_history=chat_history,
        allow_parallel_tool_calls=False,
    )

    full_response = None
    yielded_indicator = False
    async for chunk in response_stream:
        if "tool_calls" not in chunk.message.additional_kwargs:
            # Yield a boolean to indicate whether the response is a tool call
            if not yielded_indicator:
                yield False
                yielded_indicator = True

            # if not a tool call, yield the chunks!
            yield chunk  # type: ignore
        elif not yielded_indicator:
            # Yield the indicator for a tool call
            yield True
            yielded_indicator = True

        full_response = chunk

    if full_response:
        yield full_response  # type: ignore


def is_calling_different_tools(tool_calls: list[ToolSelection]) -> bool:
    """
    Check if the tool calls are from different tools.
    """
    tool_names = {tool_call.tool_name for tool_call in tool_calls}
    return len(tool_names) > 1
